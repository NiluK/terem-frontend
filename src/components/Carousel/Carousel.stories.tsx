import React from 'react';
import Carousel from './Carousel';
import Card from '../Card';

export default {
  component: Carousel,
  title: 'Carousel',
};

export const Default = () => {
  return (
    <Carousel startIndex={0} stopIndex={4} upperLimit={5} lowerLimit={0}>
      {new Array(6).fill(' ').map((val, idx) => (
        <p key={idx} data-testid={`${idx}`}>
          Hello World
        </p>
      ))}
    </Carousel>
  );
};

export const withImages = () => {
  return (
    <Carousel startIndex={0} stopIndex={4} upperLimit={5} lowerLimit={0}>
      {new Array(10).fill(' ').map((val, idx) => (
        <img key={idx} src="http://placehold.it/350x350" />
      ))}
    </Carousel>
  );
};

export const withCards = () => {
  return (
    <Carousel startIndex={0} stopIndex={4} upperLimit={5} lowerLimit={0}>
      {new Array(10).fill(' ').map((val, idx) => (
        <Card
          key={idx}
          img={'http://placehold.it/350x350'}
          imageHeight={150}
          title="Example"
          location="Example Location"
        />
      ))}
    </Carousel>
  );
};
