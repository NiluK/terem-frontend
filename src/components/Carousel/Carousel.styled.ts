import styled from 'styled-components';

export const CarouselContainer = styled.div`
  display: grid;
  position: relative;
  grid-gap: 20px;
  max-width: 100%;
  grid-template-columns: repeat(2, 1fr);
  @media screen and (min-width: 995px) {
    grid-template-columns: repeat(4, 1fr);
  }
  overflow: hidden;
`;

interface Button {
  left?: boolean;
}

export const Button = styled.button<Button>`
  position: absolute;
  top: 0;
  bottom: 0;
  border: none;
  text-align: center;
  display: flex;
  font-size: 24px;

  background: linear-gradient(
    to ${(props): string => (props.left ? 'left' : 'right')},
    rgba(255, 255, 255, 0.1) 0%,
    rgba(255, 255, 255, 1) 100%
  );
  ${(props): string => (props.left ? 'left: 0' : 'right: 0')}
`;
