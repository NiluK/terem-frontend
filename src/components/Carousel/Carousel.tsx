import React, { useState } from 'react';
import * as Styled from './Carousel.styled';

interface Carousel {
  startIndex: number;
  stopIndex: number;
  lowerLimit: number;
  upperLimit: number;
  children: JSX.Element[];
}

const Carousel = (props: Carousel): JSX.Element => {
  const { startIndex, stopIndex, upperLimit, lowerLimit, children } = props;
  const [carouselStart, setCarouselStart] = useState(startIndex);
  const [carouselStop, setCarouselStop] = useState(stopIndex);

  const onNextClick = (): void => {
    setCarouselStart(carouselStart + 1);
    setCarouselStop(carouselStop + 1);
  };

  const onPrevClick = (): void => {
    setCarouselStart(carouselStart - 1);
    setCarouselStop(carouselStop - 1);
  };

  const renderedItems = children?.slice(carouselStart, carouselStop);

  return (
    <Styled.CarouselContainer>
      {carouselStart !== lowerLimit && (
        <Styled.Button
          left
          data-testid="prev-button"
          disabled={carouselStart === lowerLimit}
          onClick={onPrevClick}
        >
          &#x3008;
        </Styled.Button>
      )}
      {renderedItems}
      {carouselStop !== upperLimit && (
        <Styled.Button
          data-testid="next-button"
          disabled={carouselStop === upperLimit}
          onClick={onNextClick}
        >
          &#x3009;
        </Styled.Button>
      )}
    </Styled.CarouselContainer>
  );
};

export default Carousel;
