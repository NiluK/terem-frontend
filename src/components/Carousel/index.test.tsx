import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import Carousel from './Carousel';

describe('Carousel Structure', () => {
  let container;
  beforeEach(() => {
    container = render(
      <Carousel startIndex={0} stopIndex={4} upperLimit={5} lowerLimit={0}>
        {new Array(6).fill(' ').map((val, idx) => (
          <p key={idx} data-testid={`${idx}`}>
            Hello World
          </p>
        ))}
      </Carousel>,
    );
  });

  it('matches snapshot', () => {
    expect(container).toMatchSnapshot();
  });

  it('renders 4 paragraphs', () => {
    expect(container.getAllByText('Hello World').length).toEqual(4);
  });

  it('renders a next button', () => {
    expect(container.getByTestId('next-button')).toBeInTheDocument();
  });

  it('renders next paragraph if the next button pressed', () => {
    const node = container.getByTestId('next-button');
    fireEvent.click(node);
    expect(container.getByTestId('4')).toBeInTheDocument();
    expect(container.getByTestId('prev-button')).toBeInTheDocument();
  });

  it('renders prev paragraph if the prev button pressed', () => {
    const next = container.getByTestId('next-button');
    fireEvent.click(next);
    const prev = container.getByTestId('prev-button');
    fireEvent.click(prev);
    expect(container.getByTestId('0')).toBeInTheDocument();
    expect(container.queryByTestId('prev-button')).toBeFalsy();
  });
});
