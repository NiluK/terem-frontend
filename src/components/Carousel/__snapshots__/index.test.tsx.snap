// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Structure matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Carouselstyled__CarouselContainer-sc-10aqm96-0 cYCoPc"
      >
        <p
          data-testid="0"
        >
          Hello World
        </p>
        <p
          data-testid="1"
        >
          Hello World
        </p>
        <p
          data-testid="2"
        >
          Hello World
        </p>
        <p
          data-testid="3"
        >
          Hello World
        </p>
        <button
          class="Carouselstyled__Button-sc-10aqm96-1 eOHtfr"
          data-testid="next-button"
        >
          〉
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Carouselstyled__CarouselContainer-sc-10aqm96-0 cYCoPc"
    >
      <p
        data-testid="0"
      >
        Hello World
      </p>
      <p
        data-testid="1"
      >
        Hello World
      </p>
      <p
        data-testid="2"
      >
        Hello World
      </p>
      <p
        data-testid="3"
      >
        Hello World
      </p>
      <button
        class="Carouselstyled__Button-sc-10aqm96-1 eOHtfr"
        data-testid="next-button"
      >
        〉
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
